socs/agents/hwp_gripper/agent.py

```diff
@@ -470,6 +525,87 @@ def run_and_append(func, *args, **kwargs):
                        check_shutdown=check_shutdown)
         time.sleep(1)
 
+        # We should stop schedule if we have an error in this task
+        if (not all(finished)) or any(aborted):
+            raise ValueError("Failed to grip HWP. Grippers are retracted.")
+
+        return data
+
+    def ungrip(self, session, params=None):
+        """ungrip()
+
+        **Task** - Series of commands to automatically ungrip the HWP.
+        This will return grippers to their home position, and retract
+        grippers as much as possible.
+        """
+
+        self._ungrip_hwp(session, check_shutdown=True)
+        return True, "Finished ungripping procedure"
+
+    def _ungrip_hwp(self, session, check_shutdown=True):
+        """
+        Helper function to ungrip the HWP that can be called by the ungrip_hwp
+        task or by the shutdown procedure.  This will return grippers to their
+        home position, and retract as much as possible.
+
+        Args
+        ------
+        session (OpSession):
+            Session object for current operation. This function will add to
+            session data.
+
+        Notes:
+            The most recent data collected is stored in session data in the
+            structure:
+
+                >>> response.session['responses']
+                [{'result': True, 'log': [.., .., etc]},
+                                    ..
+                 {'result': True, 'log': [.., .., etc]}]
```

Comment:
Thanks for including this `session.data` documentation. It should be moved up to the `ungrip()` task, otherwise it will not be visible on readthedocs.